extensions/api/observability/README.md

```diff
@@ -0,0 +1,35 @@
+# Connector Observability
+
+### `/check/health` endpoint
+
+returns information about the system status. This is primarily intended to be used
+in [`HEALTHCHECK` instructions in Docker files](https://docs.docker.com/engine/reference/builder/#healthcheck). Here,
+this endpoint always returns HTTP 200 indicating that the system is healthy as soon as this endpoint is reachable.
+
+### `/check/liveness` endpoint
+
+From the [Kubernetes docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes):
+> indicates whether the container is running.
+
+Basically the same semantics as the `/health` endpoint. We can assume, that in the event that the connector crashes, all
+REST endpoints become unavailable.
+
+### `/check/readiness` endpoint
+
+From the [Kubernetes docs](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-probes):
+> Indicates whether the container is ready to respond to requests.
+
+Thus, the readiness endpoint must return a value indicating success _only after_ all extensions that contribute an API
+have started successfully. Note, that other subsystems like the catalog crawlers might not yet be completed at that
+point, but technically the connector is "ready to respond to requests", even if they might not produce the desired - or
+even a sensible - outcome. Catalog queries may produce an empty or incomplete result until crawlers have completed.
```

Comment:
I do not fully understand why we would not wait for the container to return complete and correct results to mark it as ready? IMO the readiness endpoint should return success only when the container is ready to accept and properly process traffic (including catalog queries, etc...). Readiness probes are polled during the whole lifetime of containers, leading to Load Balancers not forwarding any traffic to them if they fail. Why would we forward traffic to containers that return incomplete results if there are others that do return correct results?

Reply:
I guess it depends what we regard as `ready to accept requests`. Why would we not accept other requests before the catalog cache is fully built? The catalog endpoint could theoretically still return a 503 until the cache is fully built.

I'd expect to be able to reach, say, my Plex server, even before its library is fully built.
Also, would triggering a new catalog crawl mean that the connector has become "unready", because its catalog may be out of date? 

I'd prefer reacting to an ongoing catalog crawl through a different mechanism. Haven't thought about it in detail, but to me, it does not contribute to readiness from an infrastructure/ops perspective

Reply:
Ok I see, assuming the catalog takes a very long time to load it might be better not to consider for infrastructure/ops readiness. But do we agree that `ready to accept requests` would include all components required for the asynchronous processing of TPs? For instance for sample 5 it would include both the Azure Storage Container as well as the S3 bucket?

Reply:
> But do we agree that ready to accept requests would include all components required for the asynchronous processing of TPs? For instance for sample 5 it would include both the Azure Storage Container as well as the S3 bucket?

yes i think so.

Reply:
@marcgs do you mean the Azure Storage Account? On the Consumer side, the connector health provider could indeed check it is accessible, but does not know about any Storage Containers (which are managed in the Transfer lifecycle). On the Provider side, the health provider doesn't know about any Storage Accounts. (Correct me if I'm wrong)

Reply:
Assuming that the readiness of a connector is determined by it's own readiness plus the readiness of all its backend services, then every extension that deals with such a backend, should implement its own checks.

For Azure Storage, the readiness/availability could be determined by listing all Containers or getting the account info, same for S3 buckets. 

There may be backend systems where such a periodic heartbeat might not be applicable, e.g. serverless functions, which may only get provisioned on demand. Such backends should be assumed "live" until proven otherwise.